### ISSUE TITLE
Tracking Issue for asm_sym

### ISSUE THREAD

**Issue by Amanieu** [rust-lang/rust#93333]
The feature gate for the issue is `#![feature(asm_sym)]`.

### Summary

This feature adds a `sym <path>` operand type to `asm!` and `global_asm!`.
- `<path>` must refer to a `fn` or `static`.
- A mangled symbol name referring to the item is substituted into the asm template string.
- The substituted string does not include any modifiers (e.g. GOT, PLT, relocations, etc).
- `<path>` is allowed to point to a `#[thread_local]` static, in which case the asm code can combine the symbol with relocations (e.g. `@plt`, `@TPOFF`) to read from thread-local data.

### Status

~~Blocked on support for `sym` in `global_asm!`.~~

**Comment by bstrie** [comment]
Is this note from #72016 still accurate?

> The way sym is handled by the compiler (in HIR & typeck) is likely to change in the future to support it in global_asm. Currently it's a Path that's evaluated in the typeck context of the function containing the asm! but that doesn't work with global_asm! which exists outside of a function body.

**Comment by Amanieu** [comment]
Yes it is.

**Comment by dylanede** [comment]
Is there any chance of splitting stabilisation of this so that `sym` in `asm!` can be stabilised first, without waiting for `global_asm!` support?

**Comment by eddyb** [comment]
> The way sym is handled by the compiler (in HIR & typeck) is likely to change in the future to support it in global_asm. Currently it's a Path that's evaluated in the typeck context of the function containing the asm! but that doesn't work with global_asm! which exists outside of a function body.

IMO the way to do this is to require an `AnonConst` (similar to the `const {...}` feature) around the `sym` path, at least in `global_asm!`. While it might be possible to whip up a typeck session without that, it seems easier to reuse the `AnonConst` functionality (similar to how we would implement `typeof(...)`).

Only downside I can think of is that it's a bit annoying that a HIR body containing a path to a `static` looks like it's reading from it (because `static` paths are place expressions instead of value expressions) - might be bypassed by "simply" not running actually running any MIR processing on it (including MIR borrowck), I'm not sure.

> which may cause subtle breaking changes (in particular with regards to functions with generic parameters).

I would not want to disallow anything that works in paths within a function, seems like an artificial limitation when there is clearly an implementation strategy that works (i.e. the current one), and `global_asm!` is generally more limited anyway (making identical implementation between the two questionable).

**Comment by Amanieu** [comment]
#94468 adds `sym` support for `global_asm!` which should help unblock the stabilization of `sym`.

**Comment by bstrie** [comment]
Now that https://github.com/rust-lang/rust/pull/94468 has landed, is all that's left to make a stabilization report?

**Comment by Amanieu** [comment]
# Stabilization report

This is a stabilization request for the `asm_sym` feature.

### Description

This feature adds a `sym <path>` operand type to `asm!` and `global_asm!`.
- `<path>` must refer to a `fn` or `static`.
- A mangled symbol name referring to the item is substituted into the asm template string.
- The substituted string does not include any modifiers (e.g. GOT, PLT, relocations, etc).
- `<path>` is allowed to point to a `#[thread_local]` static, in which case the asm code can combine the symbol with relocations (e.g. `@plt`, `@TPOFF`) to read from thread-local data.

**Comment by marcgalois** [comment]
Is the intention that generic functions will be allowed as `sym` operands (#96330) when `asm_sym` is stabilized? Or will that remain as a nightly-only feature?

**Comment by Lokathor** [comment]
I would expect a particular monomorphism of a generic function to be legal, but not that a generic function itself is legal without the generics being specified.

**Comment by marcgalois** [comment]
Does that mean that something like:

```rust
#![feature(asm_sym)]
fn func<T>() {}

fn blarg<T>() {
    unsafe {
        std::arch::asm!("/* {} */", sym func::<T>);
    }
}

fn foo() {
    blarg::<i32>();
}
```

would or wouldn't be legal? (In that `blarg` specifies a specific monomorphism of `func`, although exactly which monomorphism is determined by the type arguments to `blarg`.)

**Comment by Amanieu** [comment]
It will eventually be legal. It doesn't currently work yet due to issues with the implementation in the compiler.

**Comment by newpavlov** [comment]
It does not look like this feature currently works correctly with `#[thread_local]` statics. The following code:
```rust
#![feature(thread_local, asm_sym)]
use std::arch::global_asm;

#[thread_local]
static mut BAR: u64 = 0;

global_asm!(
    r#"
    .globl foo
    foo: mov r8, {}
    "#,
    sym BAR,
);
```

Compiles down to:
```asm
foo:
    mov     r8, qword ptr [example::BAR]

example::BAR:
    .zero   8
```
Which looks quite wrong to me. Also when code like this gets used in a binary it results in a linker error.

**Comment by bjorn3** [comment]
You have to explicitly specify the right symbol relocation kind and use the instruction sequence necessary for getting tls variables on your platform.

**Comment by newpavlov** [comment]
@bjorn3 Can't compiler insert those automatically? The problem is that the instruction sequence [may depend](https://github.com/rust-lang/rust/issues/29594#issuecomment-843790757) on compilation options and visibility of the static variable, so it could be impossible to write portable code dependent on thread local statics outside of very controlled environments.

**Comment by Amanieu** [comment]
If you want the compiler to resolve the address of a TLS variable for you then you should do so in Rust code and pass the address as an input register.

**Comment by newpavlov** [comment]
Unfortunately, it's not quite possible in my case. I place pointer to a global asm block on a custom stack, so code which gets executed on it automatically jumps to the block after simple `ret`. Inside this block I read another pointer from thread local variable and jump to it. I can not pre-compute the offset during stack creation since it may change or even get executed on a different thread from one which created the stack frame.

Is there a point to allowing `sym` to point to a thread local variable while not automatically computing offset for it? Right now you have to do it manually inside the asm code, since there are simply no other uses for it. And as I wrote earlier, it may not be possible to write portable code which performs this computation.

**Comment by roblabla** [comment]
AFAIK the correct sequence to access thread_local on x86_64 (on linux, at least) is:

```

        std::arch::asm!("
            mov r8, fs:{}@TPOFF
            mov rax, [r8]
        ", sym BAR)
```

That is, you have to split the access in two instructions: One that gets the offset, and one that actually dereferences. Trying to do it in a single instruction (something like `mov r8, [fs:{}@TPOFF]`) is invalid, because there are no instructions of the form `mov REG, [fs:IMM]`. At least, that's my understanding.

So doing this kind of rewrite would involve doing a complex rewrite of the asm in order to split the instruction, which feels like it'd probably be a lot of trouble.

**Comment by Amanieu** [comment]
You can use `-Z tls-model=local-exec` to force the use of a particular TLS model. This is useful if you are building an executable since you know all crates are going to be part of the local executable.

**Comment by bjorn3** [comment]
> AFAIK the correct sequence to access thread_local on x86_64 (on linux, at least) is:

It depends on the tls model, but for the global dynamic model, which rustc defaults to, it involves a function call, two different relocations and a bunch of padding to allow the linker to relax it to the sequence for a quicker tls model sequence if applicable. See https://github.com/bytecodealliance/wasmtime/blob/c7be93753ac9fcdff816eaccad00986a6ac06b27/cranelift/codegen/src/isa/x64/inst/emit.rs#L2840

> You can use -Z tls-model=local-exec to force the use of a particular TLS model.

That only affects LLVM generated code, not inline assembly. Also mixing several tls models for the same object format in a single object file is valid, so using `-Ztls-model` isn't necessary for this use case. It's not like rustc validates that a tls model allowed for the actual usage is used.

**Comment by newpavlov** [comment]
@bjorn3 
>No, the inline assembly text is defined as an opaque blob not parsed by the compiler.

I mean allowing compiler to insert prelude which computes address of thread local data and places it into register, which then gets used by the asm block. Something like this would be sufficient for my use-case:
```rust
#[naked]
unsafe extern "C" fn foo() {
    asm!(
        "jmp {}",
        in(reg) TLS_VAR,
    )
}
```
Instead of `in(reg)` it could be something like `sym(reg)`.

**Comment by bjorn3** [comment]
That would not be valid. For naked functions it is not know which registers need to be preserved and which can be clobbered. It is also not guaranteed that there is a valid stack pointer. This makes it impossible for the compiler to calculate the tls address in a prelude without potentially breaking things as the calculation may need to call a function like `__tls_get_addr` on ELF or call the function pointed to by a `tlv` relocation for Mach-O.

**Comment by madsmtm** [comment]
Different question: Since this doesn't include the ability to specify GOT/PLT, is that a feature request that is being tracked somewhere? Or is it already possible to specify in some other way (if so, would be nice to have that as an example somewhere)?

**Comment by bjorn3** [comment]
@madsmtm You can explicitly specify the kind of relocation you want in the inline assembly. If you choose the right kind you will get a GOT or PLT reference. For example you could do

```
asm!(
    "movq rax, [rip + {}@GOTPCREL];
    # use the address stored in rax...",
    reg("rax") _,
    sym FOO,
);
```

if you need a GOT access on x86_64.

**Comment by madsmtm** [comment]
Ah cool, that was exactly what I was looking for, thanks!

**Comment by joshtriplett** [comment]
Shall we stabilize `sym` in `asm!`?

@rfcbot merge

**Comment by joshtriplett** [comment]
Shall we stabilize sym in asm!?

@rfcbot merge

**Comment by joshtriplett** [comment]
@Amanieu It does seem unfortunate to have to write potentially target-OS-specific code to handle relocations, when you might otherwise only need to write *CPU*-specific code.

Is there some variant of `sym` we could provide that *does* do whatever relocations we would have done and provides a callable/dereferenceable address?

**Comment by Amanieu** [comment]
Inline assembly is inherently target specific. In C you would have a header file with a bunch of macros to deal with target-specific details, and [the same applies in Rust](https://github.com/Amanieu/corosensei/blob/81a1a84a1f801efe66e55412a59716947c3a2bdf/src/arch/mod.rs#L5).

If you want the compiler to handle this for you then you should use an `in` operand to load the address of a variable in a register.

**Comment by Lokathor** [comment]
I'm sympathetic to wanting assembly to be portable/reusable and so "inherently target specific" might be a little overkill, but "if you want a thing in a register then assign it to a variable and put that variable in a register" does seem like a reasonable workflow.

**Comment by newpavlov** [comment]
>Inline assembly is inherently target specific. In C you would have a header file with a bunch of macros to deal with target-specific details, and [the same applies in Rust](https://github.com/Amanieu/corosensei/blob/81a1a84a1f801efe66e55412a59716947c3a2bdf/src/arch/mod.rs#L5).

As I wrote earlier, I am not sure that it's possible to do properly. Relocation may depend on linking mode, which is impossible to `cfg` our way out.

@Lokathor 
Note that it's not possible to do in the context of `global_asm!`/naked functions.

**Comment by Lokathor** [comment]
For those you'd have to use a bunch of cfg the same as C would (apparently) use a bunch of macros.

and the cfg would have to go by a cargo feature or other expression that says what the linking mode will be.

**Comment by marcgalois** [comment]
Would it be possible to expose the code model and relocation model via `cfg`? If so, then would it be possible to write a macro to generate all the `#[cfg(...)]` statements to generate the proper assembly for getting the address of a symbol?

**Comment by joshtriplett** [comment]
@Amanieu wrote:
> Inline assembly is inherently target specific.

It's inherently *architecture*-specific; it doesn't necessarily have to be *OS*-specific.

**Comment by nikomatsakis** [comment]
@rfcbot concern rust-reference

Is there a PR adding this to the rust reference?

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by Amanieu** [comment]
No, but the addition is pretty small. It's basically a revert of [this](https://github.com/rust-lang/reference/commit/c41b359723f2459b387d690153050fbb1f3c9939#diff-8446b9df5013451106cbd2e594f27924bef78bb3b9995697c4881cbc3c6ee4a7L92-L96).

**Comment by nikomatsakis** [comment]
@Amanieu great! Can you open a PR with that revert?

**Comment by Amanieu** [comment]
https://github.com/rust-lang/reference/pull/1270

**Comment by nikomatsakis** [comment]
@rfcbot resolve rust-reference

**Comment by Amanieu** [comment]
> It sounds like all that is left is for an actual stabilization PR to be submitted?

#103168

**Comment by SUPERCILEX** [comment]
Shouldn't this issue be closed since the feature was stabilized?
